### PR TITLE
add evaluation and progress bar

### PR DESCRIPTION
**Pull request by RowanBevan**

### CODE REVIEW

File: task2.py [hunk 1]
```diff
@@ -114,6 +116,81 @@ def generate_templates(image_class, image, levels, angles, verbose=False, use_ca
     return templates
 
 
+def compute_bounding_box(match,templates):
+    emoji, result, index = match
+    ij = np.unravel_index(np.argmax(result), result.shape)
+    x, y = ij[::-1]
+    w, h = templates[emoji][index].shape
+    return ((x,y),(x+w,y+h))
```
Comment (bridges-wood) [inline task2.py:127]: Which part of the bounding box does this correspond to?

**Review by cg-2611 — APPROVED** [review]
LGTM

File: task2.py [hunk 2]
```diff
@@ -114,6 +116,81 @@ def generate_templates(image_class, image, levels, angles, verbose=False, use_ca
...
+
+    # bl is the bottom left of the given box and tr is the top right
+    bl, tr, x, y = (0, 1, 0, 1)
+
+    box_1_area = (box_1[tr][x] - box_1[bl][x]) * (box_1[tr][y] - box_1[bl][y])
+    box_2_area = (box_2[tr][x] - box_2[bl][x]) * (box_2[tr][y] - box_2[bl][y])
+
+    intersect = (min(box_1[tr][x], box_2[tr][x]) - max(box_1[bl][x], box_2[bl][x])) * (min(box_1[tr][y], box_2[tr][y]) - max(box_1[bl][y], box_2[bl][y]))
+
+    if intersect < 0:
```
Comment (bridges-wood) [inline task2.py:144]: When does this happen?

File: task2.py [hunk 3]
```diff
@@ -114,6 +116,81 @@ def generate_templates(image_class, image, levels, angles, verbose=False, use_ca
...
+
+    true_positives = 0
+    false_positives = 0
+    accuracy = []
+    # Determine if we have a: true positive, false positive for each match
+    for match in matches:
+        emoji = match[0]
+        bounding_box = compute_bounding_box(match, templates)
+
+        emoji = emoji.rsplit(".")[0].split("-",1)[1]
```
Comment (bridges-wood) [inline task2.py:179]: What is this doing?

File: task2.py [hunk 4]
```diff
@@ -125,44 +202,63 @@ def main(training_images_path, test_images_path, ground_truths_path, angles, lev
...
+    for key, test_image in tqdm(test_images.items()):
+        # Preprocess the test image
+        test_image = preprocess(test_image)
+        annotation_file_path = ground_truths_path + '/' + (key.split('.')[0])
+        if (Path(annotation_file_path + '.txt')).exists():
+            annotation_file_path += '.txt'
+        elif (Path(annotation_file_path + '.csv')).exists():
+            annotation_file_path += '.csv'
+        else:
+            print(f'No matching annotations found for test image {key}, continuing to next test image...')
```
Comment (bridges-wood) [inline task2.py]: Should this happen only when the verbose flag is set?

File: task2.py [hunk 5]
```diff
@@ -125,44 +202,63 @@ def main(training_images_path, test_images_path, ground_truths_path, angles, lev
...
+                result = match_template(test_image, template)
 
-            if verbose:
-                print(result.max())
+                if verbose:
+                    print(result.max())
 
-            if result.max() > 0.8:
-                matches.append((emoji, result, index))
+                if result.max() > 0.95:
```
Comment (bridges-wood) [inline task2.py]: We should pass this threshold as a parameter.

**Review by bridges-wood — CHANGES_REQUESTED** [review]
Couple of comments to add and one change.

**Review by bridges-wood — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**